Source/Diagnostics/ReducedDiags/ParticleMomentum.cpp

```diff
@@ -0,0 +1,262 @@
+        {
+            return p.rdata(PIdx::w);
+        });
+
+        // Reduced sum over MPI ranks
+        ParallelDescriptor::ReduceRealSum(Px  , ParallelDescriptor::IOProcessorNumber());
+        ParallelDescriptor::ReduceRealSum(Py  , ParallelDescriptor::IOProcessorNumber());
+        ParallelDescriptor::ReduceRealSum(Pz  , ParallelDescriptor::IOProcessorNumber());
+        ParallelDescriptor::ReduceRealSum(Wtot, ParallelDescriptor::IOProcessorNumber());
+
+        // Save results for this species i_s into m_data
+
+        // Offset:
+        // 3 values of total momentum for all  species +
+        // 3 values of total momentum for each species
+        offset_total_species = 3 + i_s*3;
+        m_data[offset_total_species+0] = Px;
+        m_data[offset_total_species+1] = Py;
+        m_data[offset_total_species+2] = Pz;
+
+        // Offset:
+        // 3 values of total momentum for all  species +
+        // 3 values of total momentum for each species +
+        // 3 values of mean  momentum for all  species +
+        // 3 values of mean  momentum for each species
+        offset_mean_species = 3 + nSpecies*3 + 3 + i_s*3;
+        if (Wtot > std::numeric_limits<Real>::min())
+        {
+            m_data[offset_mean_species+0] = Px / Wtot;
+            m_data[offset_mean_species+1] = Py / Wtot;
+            m_data[offset_mean_species+2] = Pz / Wtot;
+        }
+        else
+        {
+            m_data[offset_mean_species+0] = 0.0;
+            m_data[offset_mean_species+1] = 0.0;
+            m_data[offset_mean_species+2] = 0.0;
+        }
+    }
+
+    // Save total momentum and total mean momentum
+
+    m_data[0] = 0.0;
+    m_data[1] = 0.0;
+    m_data[2] = 0.0;
+
+    // Offset:
+    // 3 values of total momentum for all  species +
+    // 3 values of total momentum for each species
+    offset_mean_all = 3 + nSpecies*3;
+    m_data[offset_mean_all+0] = 0.0;
+    m_data[offset_mean_all+1] = 0.0;
+    m_data[offset_mean_all+2] = 0.0;
+
+    // Loop over species
+    for (int i_s = 0; i_s < nSpecies; ++i_s)
+    {
+        // Offset:
+        // 3 values of total momentum for all  species +
+        // 3 values of total momentum for each species
+        offset_total_species = 3 + i_s*3;
+        m_data[0] += m_data[offset_total_species+0];
+        m_data[1] += m_data[offset_total_species+1];
+        m_data[2] += m_data[offset_total_species+2];
+
+        // Offset:
+        // 3 values of total momentum for all  species +
+        // 3 values of total momentum for each species +
+        // 3 values of mean  momentum for all  species +
+        // 3 values of mean  momentum for each species
+        offset_mean_species = 3 + nSpecies*3 + 3 + i_s*3;
+        m_data[offset_mean_all+0] += m_data[offset_mean_species+0];
+        m_data[offset_mean_all+1] += m_data[offset_mean_species+1];
+        m_data[offset_mean_all+2] += m_data[offset_mean_species+2];
```

Comment:
I don't think the mean momentum for all species should be the sum of the mean momentum of individual species. (if I have two species that have the same mean momentum `p_mean`, I think the mean momentum for all species should be `p_mean`, not `2*p_mean`)
It's probably better to divide the total momentum for all species by the sum of weights over all species.
I think the same should be done for the FieldEnergy reduced diag as well.

Reply:
About this one, if you have two species both with mean momentum `p_mean` and you want what we call "mean momentum for all species" to be `p_mean` as well (which makes sense, I agree with you), shouldn't we divide simply by the number of species then? So `2*p_mean / 2 = p_mean` in the example. I don't think it would be guaranteed to give `p_mean` if we divide by the sum of weights over all species, would it? I could be missing something.

Reply:
> I don't think it would be guaranteed to give `p_mean` if we divide by the sum of weights over all species, would it?

Yes, we would get `p_mean` if we divide the _total_ momentum (i.e. what is in the first 3 columns of this diag) by the sum of all weights of both species.

When you're writing `2*p_mean / 2 = p_mean`, it looks like you're suggesting to sum the _average_ momentum of each species (and then divide by the number of species). But I don't think we should do that because that would give the same weight to species that only contain a few (or even 0) particles as to species that contain many particles.

I'm not sure how much this value makes sense anyways but I feel like it's better to output the average momentum of all the particles in the simulation than the average of the mean momentum of each species. In the latter case, simply adding a species with no particle would change the average momentum, which I don't think makes sense.

Reply:
Sounds good, I think I implemented what you suggested (total momentum divided by sum of weights over all species) in https://github.com/ECP-WarpX/WarpX/pull/1908/commits/226a96cdcf2e7a79d0350f70e328ecb70b493f6e. Let me know if this looks correct to you. If so, I will possibly change what we do for the particle energy accordingly as well.

Reply:
Yes, that looks good to me, thanks! Good idea to add a check in the analysis script! And I also agree that we should also make the change for the particle energy diag.

Reply:
Particle energy done in https://github.com/ECP-WarpX/WarpX/pull/1908/commits/94836cc028755584423d9c412ff38061badbe1af. Interestingly, there was some numerical cancellation issue causing a relative error of about 1e-7 (instead of machine precision) on the protons particle energy (which wasn't tested before). The modification
```cpp
-  return ( std::sqrt(us*c2 + c2*c2) - c2 ) * m * w;
+  return (std::sqrt(us * m2 * c2 + m2 * c4) - m * c2) * w;
```
should give a result that matches the one obtained by post-processing the yt data to machine precision. Not 100% sure about which one of the two is supposed to be more correct, from a numerical cancellation point of view, though. At least, I haven't looked into this in detail.

Reply:
Yes it's good to know that the particle energy calculation is inaccurate in the nonrelativistic limit, thanks. (although it's not surprising when looking at the formula)

If we ever need to be more accurate in this regime, we could check if the kinetic energy is much smaller than the rest mass and if so use the classical formula `p^2/2m` (or even go a bit further in the Taylor expansion). I don't think this is really an issue for now though, and even a relative error of about 1e-7 is sufficient for most purposes.

Regarding which of your 2 formulas is the more correct, I think they should have the exact same accuracy since the problematic subtraction is the same (the terms are just multiplied by `m` in the second case). Of course the second formula will give you a machine precision error in the python analysis scripts, since the exact same one that is used there, but I don't think it is more accurate per se. (actually it looks like the second formula does an extra multiplication per particle, but I don't think this should have any significant effect on the simulation speed, so we probably don't have to worry about that)

More generally, your PR looks good, thanks a lot! I just had a minor comment regarding the case where the total particle weight is 0. (see below)

Reply:
Sounds good. The fact that the two formulas should have the same accuracy, because they just differ uniformly by a multiplication, is indeed why I was a but surprised to see that they lead to a 7-digit difference in the results. I agree that it is natural to match the Python calculation to machine precision if we use the exact same formula, but I would not expect to lose/gain 7 digits by using an alternative formula that supposedly has the same accuracy (because the form of the cancellation is the same).

Anyways, thanks for the suggestion on the total particle weight, I'll add a check on that. Then I'll wait until the latest AMReX commit becomes available in WarpX, in order to try using the new tuple-level reductions for particle containers that Andrew implemented, as Axel had suggested initially.

Reply:
>  I would not expect to lose/gain 7 digits by using an alternative formula that supposedly has the same accuracy

I don't think we actually lose or gain accuracy by switching from one formula to the other. Both formulas yield results that are off by ~1e-7 (relatively) with respect to the actual mathematical result. And because the numerical error is mostly random, the two results that the formulas yield are also off with each other by ~1e-7 (which is what you saw in the first analysis script).